## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [269]:
import pandas as pd
import plotly.express as px

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

Define some utility functions.

In [270]:
# Define utility functions to calculate central tendency measures, find lower and upper bounds,
# and find outliers.

def is_number_series(data_list):
    ''' Checks whether a given data series conatins a list of numbers.

        Args:
            data_list (series of type float or int): The list of data to check

        Returns:
            True if input is a list of numbers.
            False, otherwise.
        '''
    import pandas as pd

    # Check whether input is a series.
    if isinstance(data_list, pd.Series):
        # Check whether series contains only numbers (i.e., float or int).
        # If not print an error message and return False
        for data in data_list:
            if (type(data)!=float):
                if (type(data)!=int):
                    print(f'Error: {data_list} contains data, {data}, which is not a number (int or float).')
                    return False
    else:
        # If the given data list is not a series print an error message and return False.
        print(f'Error: {data_list} is not a series. Cannot calculate mean, median, and mode.')
        return False
    
    return True

def calculate_central_tendency_measures(data_list):
    ''' Calculates the centeral tendency measures mean, median, and mode, for a series of data.

        Args:
            data_list (series of type float or int): The list of data for which to calculate the
            mean, median, and mode.

        Returns:
            A tupel of floats containing mean, median, and mode.
        '''
    import numpy as np
    import scipy.stats as sts

    # Check whether input is a series of numbers.
    if is_number_series(data_list):
        # Calculate mean, median, and mode.
        mean = np.mean(data_list)
        median = np.median(data_list)
        mode = sts.mode(data_list)
    else:
        print(f'Error: {data_list} is not a series of numbers. Cannot calculate mean, median, and mode.')
        mean = float('nan')
        median = float('nan')
        mode = float('nan')

    return (mean, median, mode)

def find_outliers(data_list):
    ''' Calculates the outliers for a series of data.

        Args:
            data_list (series of type float or int): The list of data for which to calculate the
            outliers.

        Returns:
            A series of outliers.
        '''
    
    # Check whether input is a series of numbers.
    if is_number_series(data_list):
        # Calculate upper and lowere bound
        (lower_bound, upper_bound) = find_lower_upper_bounds(data_list)

        # Calculate outliers
        outliers = data_list.loc[(data_list < lower_bound) | (data_list > upper_bound)]
    else:
        print(f'Error: {data_list} is not a series of numbers. Cannot calculate outliers.')
        outliers = float('nan')

    return outliers

def find_lower_upper_bounds(data_list):
    ''' Calculates the lower and upper bounds for a series of data.

        Args:
            data_list (series of type float or int): The list of data for which to calculate the
            outliers.

        Returns:
            A tupel conatining lower bound and upper bound.
        '''
    
    # Check whether input is a series of numbers.
    if is_number_series(data_list):
        # Calculate inter quartile range, IQR:
        Q1 = data_list.quantile(0.25)
        Q3 = data_list.quantile(0.75)
        IQR = Q3 - Q1

        # Calculate lower and upper bound.
        lower_bound = Q1 - (1.5 * IQR)
        upper_bound = Q3 + (1.5 * IQR)
    else:
        print(f'Error: {data_list} is not a series of numbers. Cannot calculate lower and upper bounds.')
        lower_bound = float('nan')
        upper_bound = float('nan')

    return (lower_bound, upper_bound)

In [271]:
# Clean the data if necessary.
# Get the number of rows and columns
(rows, columns) = df.shape
print(f"The number of rows is {rows}.")
print(f"The number of columns is {columns}.")

# Get the number of records and compare it with the number of rows obtained before.
# They should be equal.
records_count = max(df.count())
if records_count == rows:
    print(f"The number of records, {records_count}, is equal to the number of rows, {rows}.")
else:
    print(f"Data cleaning required since the number of records, {records_count}, differes from the number of rows, {rows}.")

# Check for 'null' values
df.isnull().head(10)

The number of rows is 54639.
The number of columns is 18.
The number of records, 54639, is equal to the number of rows, 54639.


first   last    job  phone  email  client_id  order_id  order_date  \
0  False  False  False  False  False      False     False       False   
1  False  False  False  False  False      False     False       False   
2  False  False  False  False  False      False     False       False   
3  False  False  False  False  False      False     False       False   
4  False  False  False  False  False      False     False       False   
5  False  False  False  False  False      False     False       False   
6  False  False  False  False  False      False     False       False   
7  False  False  False  False  False      False     False       False   
8  False  False  False  False  False      False     False       False   
9  False  False  False  False  False      False     False       False   

   order_week  order_year  item_id  category  subcategory  unit_price  \
0       False       False    False     False        False       False   
1       False       False    False     False        False       False   
2       False       False    False     False        False       False   
3       False       False    False     False        False       False   
4       False       False    False     False        False       False   
5       False       False    False     False        False       False   
6       False       False    False     False        False       False   
7       False       False    False     False        False       False   
8       False       False    False     False        False       False   
9       False       False    False     False        False       False   

   unit_cost  unit_weight    qty  line_number  
0      False        False  False        False  
1      False        False  False        False  
2      False        False  False        False  
3      False        False  False        False  
4      False        False  False        False  
5      False        False  False        False  
6      False        False  False        False  
7      False        False  False        False  
8      False        False  False        False  
9      False        False  False        False

There do not seem to be any null values.
Just in case create a new dataframe without any null values and compare the number of records with the 
original datafarme.

In [272]:
# Create new dataframe without any null values.
df_no_nulls = df.dropna(how='any')
records_count_no_nulls = max(df_no_nulls.count())

# Check whether the number of records are equal.
if records_count_no_nulls == records_count:
    print(f"The number of records before and after removal of null values is the same, {records_count_no_nulls}.")
else:
    print(f'{records_count-records_count_no_nulls} records were dropped. Please make sure the data lost will not affect the following analysis.')

The number of records before and after removal of null values is the same, 54639.


No records were dropped, i.e., all original data can be used.

In [273]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [274]:
# Use the describe function to gather some basic statistics for columns where that might make sense.
df[['unit_price', 'unit_cost', 'unit_weight', 'qty']].describe()

unit_price     unit_cost   unit_weight           qty
count  54639.000000  54639.000000  54639.000000  5.463900e+04
mean     136.267207     99.446073      5.004116  5.702646e+02
std      183.873135    133.164267      5.326599  1.879552e+04
min        0.010000      0.010000      0.000000  0.000000e+00
25%       20.800000     14.840000      1.450000  3.200000e+01
50%       68.310000     49.890000      3.240000  6.800000e+01
75%      173.160000    125.570000      6.890000  1.700000e+02
max     1396.230000    846.270000     46.430000  3.958244e+06

In [275]:
# Use this space to do any additional research
# and familiarize yourself with the data.

# Compare the mean, median, and mode for 'unit_price'
(mean_unit_price, median_unit_price, mode_unit_price) = calculate_central_tendency_measures(df['unit_price'])
print(f"The mean for 'unit price' is ${mean_unit_price:.2f}.")
print((f"The median for 'unit price' is ${median_unit_price}."))
print((f"The mode for 'unit price' is ${mode_unit_price[0]}. It occurs {mode_unit_price[1]} times."))


The mean for 'unit price' is $136.27.
The median for 'unit price' is $68.31.
The mode for 'unit price' is $1.49. It occurs 249 times.


In [276]:
# Plot a histogram to get a visual of the distribution of unit price
hist = px.histogram(df, x = 'unit_price')
hist.show()

In [277]:
# Find outliers for the unit price.
unit_price_outliers = find_outliers(df['unit_price'])
unit_price_outliers.head()

0      1096.80
10      520.00
31      410.17
95     1096.80
105     448.58
Name: unit_price, dtype: float64

In [278]:
# Compare the mean, median, and mode for 'qty'
(mean_qty, median_qty, mode_qty) = calculate_central_tendency_measures(df['qty'])
print(f"The mean for 'qty' is {mean_qty}.")
print(f"The median for 'qty' is {median_qty}.")
print(f"The mode for 'qty' is {mode_qty[0]}. It occurs {mode_qty[1]} times.")

# Find maximum qty.
max_qty = max(df['qty'])
print(f"The maximum 'qty' is {max_qty:,}.")

The mean for 'qty' is 570.2646095279928.
The median for 'qty' is 68.0.
The mode for 'qty' is 24. It occurs 607 times.
The maximum 'qty' is 3,958,244.


In [279]:
# Plot a histogram to get a visual on the distribution of 'qty'.
hist = px.histogram(df, x = 'qty')
hist.show()

Because of the large maximum `qty` ordered the histogram "is bunched up" on the small values. Let's find upper and lower bounds and plot only the data within those limits.

In [280]:
# Find lower and upper bounds.
(lower_bound_qty, upper_bound_qty) = find_lower_upper_bounds(df['qty'])

# Ignore outliers
qty_no_outliers = (df['qty'].loc[(df['qty'] > lower_bound_qty) & (df['qty'] < upper_bound_qty)])

# Histogram without outliers
hist = px.histogram(qty_no_outliers, x = 'qty')
hist.show()

The shape of this distribution is similar to the shape of the `unit_price` distribution for the same reason.

In [281]:
# Find outliers
qty_outliers = find_outliers(df['qty'])
display(qty_outliers.head(10))

# Find maximum of outliers
qty_outliers.max()

12        568
24       2163
25     129944
54        804
76        802
79      29073
84        392
98       1167
106       897
109       466
Name: qty, dtype: int64

3958244

In [282]:
# What three item categories had the most entries?
# The function 'value_counts()' returns a list in descending order. Hence, we can use the head function.
top_3_categories = df['category'].value_counts().head(3)
print("The item categories with the most entries are:")
for key in top_3_categories.keys():
    print(f"   {key.title()} with {top_3_categories[key]:,} entries.")

The item categories with the most entries are:
   Consumables with 23,538 entries.
   Furniture with 11,915 entries.
   Software with 8,400 entries.


Note that in the cell above and in the following we assume that entries refers to records in the dataframe.

In [283]:
# For the category with the most entries,
# which subcategory had the most entries?
top_category = top_3_categories.keys()[0]
top_subcategory_dict = df[df['category'] == top_category]['subcategory'].value_counts().head(1)
top_subcategory = top_subcategory_dict.keys()[0]
top_subcategory_entries = top_subcategory_dict[top_subcategory]
print(f"For top category '{top_category.title()}' the top subcategory is '{top_subcategory.title()}' with {top_subcategory_entries:,} entries.")

For top category 'Consumables' the top subcategory is 'Bathroom Supplies' with 6,424 entries.


In [284]:
# Which five clients had the most entries in the data?
# Grab the client_ids with the top five number of entries
top_5_client_ids = df['client_id'].value_counts().head()

# Build a list of the five clients with the top 5 entries.
# Each element in the list is a client dictionary. The client dictionary contains
# the client ID, their full name and the number of entries.
# initialize top five client list.
top_5_clients = []
# Loop through each of the top five client ids.
for id in top_5_client_ids.index:
    # Initialize the client dictionary.
    client= {}

    # Grab first and last name from the dataframe.
    # Note that there are many entries with the same client ID. We only need the
    # first entry since the first and last names will all be the same.
    first_name = df[df['client_id'] == id].iloc[0]['first']
    last_name = df[df['client_id'] == id].iloc[0]['last']

    # Assign the values to the client dictionary keys
    client["Client ID"] = id
    client["Name"] = first_name.title() + ' ' + last_name.title()
    client["Entries"] = top_5_client_ids[id]

    # Append the client dictionary to the top_5_client list.
    top_5_clients.append(client)

print("The top five clients by number of entries are:")
for client in top_5_clients:
    print(f"   {client['Name']} (ID {client['Client ID']}) with {client['Entries']} entries.")

The top five clients by number of entries are:
   Jessica Reyes (ID 33615) with 220 entries.
   Angela Everett (ID 66037) with 211 entries.
   Bryan Myers (ID 46820) with 209 entries.
   Alexandra Young (ID 38378) with 207 entries.
   Kendra Garrett (ID 24741) with 207 entries.


In [285]:
# Store the client ids of those top 5 clients in a list.

# Already done. See above.

In [286]:
# How many total units (the qty column) did the
# client with the most entries order?
# Retrieve the top client ID from the list of top five clients
top_client_id = top_5_clients[0]['Client ID']

# Sum over 'qty' for all entries with the top client ID
total_units_top_client = df[df['client_id'] == top_client_id]['qty'].sum()

print(f"The client with the most orders, {top_5_clients[0]['Name']}, ordered {total_units_top_client:,} units.")


The client with the most orders, Jessica Reyes, ordered 64,313 units.


That does not mean they ordered the most units.

In [288]:
# Find the client that ordered the most units.
most_units = 0
for id in top_5_client_ids.index:
    total = df[df['client_id'] == id]['qty'].sum()
    if(total > most_units):
        most_units = total
        most_units_client_id = id

# Get client name
for client in top_5_clients:
    if client['Client ID'] == most_units_client_id:
        most_units_client_name = client['Name']
        break

print(f'{most_units_client_name} with client ID {most_units_client_id} ordered the most units. They ordered {most_units:,} units.')


Kendra Garrett with client ID 24741 ordered the most units. They ordered 239,862 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [289]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [290]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
df['total_weight'] = df['unit_weight'] * df['qty']

# Now calculate the shipping price using the formula given above. Add it as a new column as well.
df['shipping_price'] = df['total_weight'].apply(lambda weight: weight*7 if weight > 50 else weight*10)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  total_weight  \
0     762.71         7.50  105            1  115164.00        787.50   
1      15.09         1.49   21            0     523.95         31.29   
2       7.86         1.68   39            6     527.28         65.52   
3      24.85         1.23   29            3    1056.18         35.67   
4     108.17        46.43   20            1    3902.00        928.60   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20  

[5 rows x 21 columns]

In [291]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

# Define a function to calculate the total price using the formula given above.
def total_price(price, shipping):
    return (price + shipping)*(1+9.25/100)

# Create a new column called 'total_price'
df['total_price'] = total_price(df['subtotal'], df['shipping_price'])
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  total_weight  shipping_price  \
0         7.50  105            1  115164.00        787.50         5512.50   
1         1.49   21            0     523.95         31.29          312.90   
2         1.68   39            6     527.28         65.52          458.64   
3         1.23   29            3    1056.18         35.67          356.70   
4        46.43   20            1    3902.00        928.60         6500.20   

     total_price  
0  131839.076250  
1     914.258625  
2    1077.117600  
3    1543.571400  
4   11364.403500  

[5 rows x 22 columns]

In [292]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  total_weight  shipping_price    total_price  \
0            1  115164.00        787.50         5512.50  131839.076250   
1            0     523.95         31.29          312.90     914.258625   
2            6     527.28         65.52          458.64    1077.117600   
3            3    1056.18         35.67          356.70    1543.571400   
4            1    3902.00        928.60         6500.20   11364.403500   

       cost  
0  85597.05  
1    629.79  
2    765.18  
3   1077.35  
4   8663.60  

[5 rows x 23 columns]

In [293]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['total_price'] - df['cost']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

    subtotal  total_weight  shipping_price    total_price      cost  \
0  115164.00        787.50         5512.50  131839.076250  85597.05   
1     523.95         31.29          312.90     914.258625    629.79   
2     527.28         65.52          458.64    1077.117600    765.18   
3    1056.18         35.67          356.70    1543.571400   1077.35   
4    3902.00        928.60         6500.20   11364.403500   8663.60   

         profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [294]:
# Check your work using the totals above

# Create a list for the test order IDs.
test_order_ids = [{'Order ID': 2742071, 'Receipt': 152811.89},
                  {'Order ID': 2173913, 'Receipt': 162388.71},
                  {'Order ID': 6128929, 'Receipt': 923441.25}]

# Calculate the total prices for each test order ID and print.
for id in test_order_ids:
    # Calculate total price
    sum = df[df['order_id'] == id['Order ID']]['total_price'].sum()

    # Compare with recipt amount.
    diff = sum - id['Receipt']

    print(f"The total price for order {id['Order ID']} is ${sum:,.2f}.")
    print(f"The difference between this total price and the receipt amount is ${diff:,.2f}.")

    if abs(diff) <= 0.01:
        print("the difference is less or equal to $0.01 which is a rounding error.")
    else:
        print(f"Error: The total price for order ID {id['Order ID']} does not match the receipt amount!")
        break

The total price for order 2742071 is $152,811.90.
The difference between this total price and the receipt amount is $0.01.
the difference is less or equal to $0.01 which is a rounding error.
The total price for order 2173913 is $162,388.72.
The difference between this total price and the receipt amount is $0.01.
the difference is less or equal to $0.01 which is a rounding error.
The total price for order 6128929 is $923,441.24.
The difference between this total price and the receipt amount is $-0.01.
the difference is less or equal to $0.01 which is a rounding error.


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [295]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

# Loop through each of the top five clients and calculate their total spend.
for client in top_5_clients:
    client_spend = df[df['client_id'] == client['Client ID']]['total_price'].sum()
    print(f"The client {client['Name']} (ID {client['Client ID']}) spent ${client_spend:,.2f}")


The client Jessica Reyes (ID 33615) spent $8,377,308.57
The client Angela Everett (ID 66037) spent $10,259,514.80
The client Bryan Myers (ID 46820) spent $9,743,794.32
The client Alexandra Young (ID 38378) spent $12,906,550.88
The client Kendra Garrett (ID 24741) spent $82,268,891.98


In [296]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

# Create a dataframe with the rows and columns we need
data_needed_df = df[df.client_id.isin(top_5_client_ids.index)][['client_id', 'qty', 'shipping_price', 'total_price', 'profit']]

# Calculate the totals
totals_df = data_needed_df.groupby('client_id')[['qty', 'shipping_price', 'total_price', 'profit']].sum()

# Reset index.
reset_df = totals_df.reset_index()

#Retrieve the client names in the right order.
names = []
for id in reset_df['client_id']:
    for client in top_5_clients:
        if client['Client ID'] == id:
            names.append(client['Name'])

# Add a 'Name' column. They are already in the right order.
reset_df['Name'] = names

# Sort by profit and reorder columns
reset_sorted_df = reset_df.sort_values('profit', ascending=False)[['client_id', 'Name', 'qty', 'shipping_price', 'total_price', 'profit']]

# Divide the columns containing dollar amounts by 1,000,000 to get the amounts in
# millions of dollars (mill. of $)
reset_sorted_df[['shipping_price', 'total_price', 'profit']] = reset_sorted_df[['shipping_price', 'total_price', 'profit']]/1000000
reset_sorted_df


client_id             Name     qty  shipping_price  total_price     profit
0      24741   Kendra Garrett  239862        5.126448    82.268892  36.579992
2      38378  Alexandra Young   73667        3.429455    12.906551   3.271830
4      66037   Angela Everett   43018        1.395152    10.259515   3.255032
3      46820      Bryan Myers   75768        1.601449     9.743794   2.736603
1      33615    Jessica Reyes   64313        1.828985     8.377309   2.201995

In [304]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

# Rename columns
summary_df = reset_sorted_df.rename(columns = {'client_id': 'Customer ID',
                                               'Name': 'Customer Name',
                                               'qty': 'Total Units Ordered',
                                               'shipping_price': 'Total Shipping Cost (mill. $)',
                                               'total_price': 'Total Revenue (mill. $)',
                                               'profit': 'Total Profit (mill. $)'}).round(2)

summary_df


Customer ID    Customer Name  Total Units Ordered  \
0        24741   Kendra Garrett               239862   
2        38378  Alexandra Young                73667   
4        66037   Angela Everett                43018   
3        46820      Bryan Myers                75768   
1        33615    Jessica Reyes                64313   

   Total Shipping Cost (mill. $)  Total Revenue (mill. $)  \
0                           5.13                    82.27   
2                           3.43                    12.91   
4                           1.40                    10.26   
3                           1.60                     9.74   
1                           1.83                     8.38   

   Total Profit (mill. $)  
0                   36.58  
2                    3.27  
4                    3.26  
3                    2.74  
1                    2.20

In [298]:
# Sort the updated data by "Total Profit" form highest to lowest

# Already done. See above.

### Preliminary Conclusion

Out of the top five customers by orders placed, the customer with ID 33615 (Jessica Reyes) placed the most orders. However, they did not generate the highest profit or revenue. The customer with ID 24741 (Kendra Garrett) generated by far the most revenue and profit, although they placed fewer orders than almost everyone else out of the top five customers and had the highest shipping cost. However, they ordered the most units in one order. That leads to the hypotheses that the total number of units ordered per order is a better predictor for which customer provided the most total revenue and total profit than the number of orders placed.

To test this hypothesis, we will determine total revenue and total profits for the customer that ordered the most units in one order.


In [300]:
# Determine the client ID of the customer who ordered the most units. We determined how many units that customer ordered above when we found
# the outliers for the 'qty' distribution. They ordered 3,958,244 units.
max_units_client_id = df['client_id'].loc[df['qty'] == max_qty].values[0]
print(f'Client ID of the client that ordered the most units in one order: {max_units_client_id}.')

# Find the name of this customer.
max_units_name = str(df['first'].loc[df['client_id'] == max_units_client_id].values[0]) + ' ' + str(df['last'].loc[df['client_id'] == max_units_client_id].values[0])
print(f'Name of the client that ordered the most units in one order: {max_units_name}.')

Client ID of the client that ordered the most units in one order: 78965.
Name of the client that ordered the most units in one order: Julie Anderson.


In [303]:
# Find orders from only the customer that ordered the most units in one order
max_units_df = df[df['client_id'] == max_units_client_id][['client_id', 'qty', 'shipping_price', 'total_price', 'profit']]

# Find the totals for units ordered shipping price, total price, and profit in millions of dollars
max_units_totals_df = max_units_df.groupby('client_id')[['qty', 'shipping_price', 'total_price', 'profit']].sum()/1000000

# Add a 'Names' column
max_units_totals_df['Customer Name'] = max_units_name

# Reorder the columns
max_units_totals_df = max_units_totals_df[['Customer Name', 'qty', 'shipping_price', 'total_price', 'profit']]

# Reset the index
max_units_totals_df = max_units_totals_df.reset_index()

# Rename the columns
max_units_summary_df = max_units_totals_df.rename(columns = {'client_id': 'Customer ID',
                                                             'Name': 'Customer Name',
                                                            'qty': 'Total Units Ordered',
                                                            'shipping_price': 'Total Shipping Cost (mill. $)',
                                                            'total_price': 'Total Revenue (mill. $)',
                                                            'profit': 'Total Profit (mill. $)'}).round(2)
max_units_summary_df



Customer ID   Customer Name  Total Units Ordered  \
0        78965  Julie Anderson                 4.03   

   Total Shipping Cost (mill. $)  Total Revenue (mill. $)  \
0                           5.34                   655.92   

   Total Profit (mill. $)  
0                  247.35

As you can see the customer who ordered the most units in one order provided much more total revenue and total profit than one of the customers in the top five list of customers that placed the most orders. Therefore, for the given data set, the number of units ordered in one order is a better predictor for the total revenue and the total profit derived from a customer.

Finally, we will determine what the maximum total revenue and total profit is and whether that is the same as the total revenue and total profit derived from the customer that ordered the most units in one order.

In [302]:
# Find the sums of total price and profit for all clients.
max_totals_df = df.groupby('client_id')[['total_price', 'profit']].sum()/1000000

# Find the maximum revenue and print it.
max_revenue = max_totals_df['total_price'].max()
print(f'The maximum total revenue is ${max_revenue:.2f} million dollars.')

# Find the maximum profit and print it.
max_profit = max_totals_df['profit'].max()
print(f'The maximum total profit is ${max_profit:.2f} million dollars.')

The maximum total revenue is $655.92 million dollars.
The maximum total profit is $247.35 million dollars.


### Final Conclusion

For the given dataset the number of orders placed by a customer is not a good predictor of the total revenue and profit derived from a customer. A much better predictor is the number of units placed in one order. In fact, the customer that placed the order with the most units also turns out to be the customer with the highest overall total revenue and profit.